### DISCUSSION TITLE
interp: Make it clearer that commands won't exit/fail unless for the last command?

### DISCUSSION THREAD

**Discussion by andreynering** [mvdan/sh#649] [Ideas]
This is a feedback from the [Task](https://github.com/go-task/task) project (you may be tired of hearing of this project already 🙂).

We're now running `set -e` automatically to avoid surprises: https://github.com/go-task/task/commit/ac8e3441736a9cab4c64f8feb67619bfda17a08a.

I think you should consider:

- Making the documentation clearer that commands won't exit by default, unless the last command in the file, and that errors won't be returned at all, again, unless for the last command.
- Adding an option so `set -e` is considered by default, and maybe the main examples in the README and doc should use that.

I know you mimicked how running real shell files work, but it can still be forgotten and confusing when debugging stuff.

**Comment by mvdan** [comment]
You're right that the logic is to follow how a real shell works. We can't change the default behavior without breaking users, as you can imagine. I agree that the default is perhaps a bit surprising from the point of view of Go, since in general Go APIs that return an error also stop at the first error they encounter.

I lean towards documenting this better, because using `set -e` is already possible and rather easy. You just need to use the `Params("-e")` option, so you don't even need to parse/interpret a `set` command. Perhaps an example to complement the docs would be useful, I'm happy to review patches.

**Reply by andreynering** [reply]
Nice to know there's an option for it already
